File: Swift/677.键值映射.swift
/*
 * @lc app=leetcode.cn id=677 lang=swift
 *
 * [677] 键值映射
 */

// @lc code=start

class TrieNode {
    var val = 0
    var children: [TrieNode?] = .init(repeating: nil, count: 26)
}

class MapSum {
    
    private let root = TrieNode()
    private var cnt = [String: Int]()

    init() {

    }
    
    func insert(_ key: String, _ val: Int) {
        var delta = val
        if let val_ = cnt[key] {
            delta -= val_
        }
        cnt[key] = val
        var node = root
        for c in key.unicodeScalars {
            let node_ = node.children[Int(c.value) - 97] ?? TrieNode()
            node.children[Int(c.value) - 97] = node_
            node = node_
            node.val += delta
        }
    }
    
    func sum(_ prefix: String) -> Int {
        var node = root
        for c in prefix.unicodeScalars {
            if let node_ = node.children[Int(c.value) - 97] {
                node = node_
            } else {
                return 0
            }
        }
        return node.val;
    }
}

/**
 * Your MapSum object will be instantiated and called as such:
 * let obj = MapSum()
 * obj.insert(key, val)
 * let ret_2: Int = obj.sum(prefix)
 */
// @lc code=end

func main() {
    /**
     输入：
     ["MapSum", "insert", "sum", "insert", "sum"]
     [[], ["apple", 3], ["ap"], ["app", 2], ["ap"]]
     输出：
     [null, null, 3, null, 5]

     解释：
     MapSum mapSum = new MapSum();
     mapSum.insert("apple", 3);
     mapSum.sum("ap");           // return 3 (apple = 3)
     mapSum.insert("app", 2);
     mapSum.sum("ap");           // return 5 (apple + app = 3 + 2 = 5)
     */
    let obj = MapSum()
    obj.insert("apple", 3)
    assert(3 == obj.sum("ap"))
    obj.insert("app", 2)
    assert(5 == obj.sum("ap"))
}

File: Cpp/677.键值映射.cpp
/*
 * @lc app=leetcode.cn id=677 lang=cpp
 *
 * [677] 键值映射
 */

#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

// @lc code=start
struct TrieNode {
    int val;
    TrieNode * children[26];
    TrieNode() {
        this->val = 0;
        for (int i = 0; i < 26; ++i) {
            this->children[i] = nullptr;
        }
    }
};

class MapSum {
private:
    TrieNode *root;
    unordered_map<string, int> cnt;
public:
    MapSum() {
        this->root = new TrieNode();
    }
    
    void insert(string key, int val) {
        int delta = val;
        if (cnt.count(key)) {
            delta -= cnt[key];
        }
        cnt[key] = val;
        TrieNode *node = root;
        for (auto c : key) {
            if (node->children[c - 'a'] == NULL) {
                node->children[c - 'a'] = new TrieNode();
            }
            node = node->children[c - 'a'];
            node->val += delta;
        }
    }
    
    int sum(string prefix) {
        TrieNode *node = root;
        for (auto c : prefix) {
            if (node->children[c - 'a'] == NULL) {
                return 0;
            }
            node = node->children[c - 'a'];
        }
        return node->val;
    }
};

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum* obj = new MapSum();
 * obj->insert(key,val);
 * int param_2 = obj->sum(prefix);
 */
// @lc code=end


int main(int argc, const char * argv[]) {
    // insert code here...
    
    /**
     输入：
     ["MapSum", "insert", "sum", "insert", "sum"]
     [[], ["apple", 3], ["ap"], ["app", 2], ["ap"]]
     输出：
     [null, null, 3, null, 5]

     解释：
     MapSum mapSum = new MapSum();
     mapSum.insert("apple", 3);
     mapSum.sum("ap");           // return 3 (apple = 3)
     mapSum.insert("app", 2);
     mapSum.sum("ap");           // return 5 (apple + app = 3 + 2 = 5)
     */
    MapSum* obj = new MapSum();
    obj->insert("apple", 3);
    assert(3 == obj->sum("ap"));
    obj->insert("app", 2);
    assert(5 == obj->sum("ap"));

    cout << "OK~" << endl;

    return 0;
}
